### PR TITLE
MANTA-5069 MVP GCv2 garbage-collector components

### PR DESCRIPTION
**Pull request by joshwilsdon**

### CODE REVIEW

File: lib/garbage-buckets-consumer.js [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+        }
+
+        self.log.debug(
+            {
+                batchId: opts.batchId,
+                deleteBatchResult: deleteBatchResult,
+                elapsed: elapsedSince(beginning),
+                reqId: opts.reqId
+            },
+            'Called deleteGarbageBatch.'
```
Comment (trentm) [inline lib/garbage-buckets-consumer.js]:
```suggestion
            'Called deleteGCBatch.'
```

File: lib/garbage-uploader.js [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+//
+// This contains the code for GarbageUploader objects which are used by the
+// garbage-buckets-consumer program to process garbage collector instructions
```
Comment (trentm) [inline lib/garbage-uploader.js]:
```suggestion
// garbage-uploader program to process garbage collector instructions
```

File: Makefile [hunk 3]
```diff
@@ -208,6 +94,23 @@ publish: release
...
+
+.PHONY: fmt
+fmt: | $(ESLINT)
+	$(ESLINT) --fix $(ESLINT_FILES)
+
+.PHONY: test
+test:
+	@echo "To run tests, run:"
+	@echo ""
+	@echo "    ./bin/node test/*.test.js"
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply Makefile]: good point, thanks!

File: Makefile [hunk 4]
```diff
@@ -5,193 +5,79 @@
...
+	cp $(ROOT)/build/node/bin/node $(PROTO)$(PREFIX)/bin/
+	chmod 755 $(PROTO)$(PREFIX)/bin/node
+	mkdir -p $(PROTO)$(PREFIX)/scripts
+	cp $(ROOT)/deps/manta-scripts/*.sh $(PROTO)$(PREFIX)/scripts/
+	cp $(ROOT)/boot/*.sh $(PROTO)$(PREFIX)/scripts/
+	chmod 755 $(PROTO)$(PREFIX)/scripts/*.sh
+	(cd $(PROTO)$(PREFIX)/../boot && ln -s ../$(NAME)/scripts/{setup,configure}.sh .)
 
 .PHONY: release
 release: install
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply Makefile]: sold. I had missed this.

File: README.md [hunk 5]
```diff
@@ -28,346 +30,26 @@ details on major Manta versions.
...
-	"record_read_sort_order": "asc|ASC|desc|DESC",
-	"record_delete_batch_size": integer,
-	"record_delete_delay": integer (ms)
-}
+It's the job of the `garbage-dir-consumer` in the `garbage-collector` zone to
+pull records from Moray and write them to local "instructions" files. These
+instructions files are then sent to the "mako" (storage zone) that holds the
+actual files for the deleted objects. The `garbage-deleter` on the storage zone
+reads these instructions and does the actual deletion before deleting each
+instruction file.
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply README.md]: yep, good point.

File: README.md [hunk 6]
```diff
@@ -28,346 +30,26 @@ details on major Manta versions.
 
 # Overview
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply README.md]: Thanks.

File: README.md [hunk 7]
```diff
@@ -28,346 +30,26 @@ details on major Manta versions.
...
-```
-
-Retrieve performance tunables:
-```
-GET /tunables
+This repo contains the bulk of the code for the Manta v2 garbage collector
+system. At a high-level, this system looks like:
 
-Response:
+![GCv2 Overview](img/GCv2-overview.png)
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply README.md]: indeed.

File: boot/configure.sh [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+#
+
+#
+# Runs on every boot of a "garbage-collector" zone.
+#
+# (Installed as "configure.sh" which is the standard name, and will be executed
+# by the "user-script")
+#
+
+printf '==> configure.sh @ %s\n' "$(date -u +%FT%TZ)"
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply boot/configure.sh]: I think I'll leave it just so if someone in the future wants to use it, they don't accidentally create the file with the wrong name and wonder why it's not working.

File: boot/setup.sh [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+set -o xtrace
+
+NAME=manta-garbage-collector
+
+#
+# Once each time we (re)provision the "garbage-collector" zone this script will
+# run and perform the setup functions.
+#
+# (Installed as "setup.sh" which is the standard name, and this will be executed
+# by the "user-script")
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply boot/setup.sh]: ok.

File: boot/setup.sh [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+SPOOL_DIR="/var/spool/manta_gc"
+SVC_ROOT="/opt/smartdc/$NAME"
+
+#
+# Build PATH from this list of directories.  This PATH will be used both in the
+# execution of this script, as well as in the root user .bashrc file.
+#
+paths=(
+    "$SVC_ROOT/bin"
+    "$SVC_ROOT/node/bin"
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply boot/setup.sh]: Nice catch.

File: boot/setup.sh [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+#
+if ! /usr/bin/ed -s '/root/.bashrc'; then
+    fatal 'could not modify .bashrc'
+fi <<EDSCRIPT
+/export PATH/d
+a
+export PATH="$PATH"
+.
+w
+EDSCRIPT
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply boot/setup.sh]: wfm. thanks.

File: boot/setup.sh [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+if ! source "$SVC_ROOT/scripts/util.sh" ||
+    ! source "$SVC_ROOT/scripts/services.sh"; then
+        exit 1
+fi
+
+manta_common_presetup
+
+manta_add_manifest_dir "/opt/smartdc/$NAME"
+
+manta_common_setup 'garbage-collector'
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply boot/setup.sh]: Which repo do I look at to get a copy of the manta_common2_setup functions? Those don't exist in the zone currently. Also is that really the name? `common2`? 😂

[the remark this answers is not in the document]
Comment (joshwilsdon) [reply boot/setup.sh]: Ah, I guess I have to update `manta-scripts`.

File: docs/basic.md [hunk 13]
```diff
@@ -1,5 +1,7 @@
 # Basic Functionality Demo
 
+XXX 2019-10-28 A lot of this is no longer relevant for Manta v2. Needs to be rewritten.
+
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply docs/basic.md]: Sure.

File: smf/manifests/garbage-buckets-consumer.xml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+        <dependency name="dep0" grouping="require_all" restart_on="error" type="service">
+            <service_fmri value="svc:/milestone/multi-user:default" />
+        </dependency>
+
+        <exec_method type="method" name="start" timeout_seconds="10" exec="/opt/smartdc/manta-garbage-collector/bin/node --abort-on-uncaught-exception /opt/smartdc/manta-garbage-collector/bin/garbage-buckets-consumer.js &amp;">
+            <method_context>
+                <method_environment>
+                    <envvar name="HOME" value="/root" />
+                    <envvar name="PATH" value="/opt/smartdc/manta-garbage-collector/bin:/opt/smartdc/manta-garbage-collector/node/bin:/opt/local/bin:/opt/local/sbin:/usr/sbin:/usr/bin:/sbin" />
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply smf/manifests/garbage-buckets-consumer.xml]: Oops.

File: lib/common.js [hunk 15]
```diff
@@ -5,82 +5,621 @@
...
 
+function createLogger(opts) {
+    assert.object(opts, 'opts');
+    assert.string(opts.name, 'opts.name');
 
-var MORAY_CONNECT_TIMEOUT = 10000;
+    var logger = bunyan.createLogger({
+        level: opts.level || process.env.LOG_LEVEL || bunyan.INFO,
+        name: opts.name,
+        serializers: bunyan.stdSerializers
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/common.js]: ok

File: lib/common.js [hunk 16]
```diff
@@ -5,82 +5,621 @@
...
+        );
+
+        assert.uuid(config.instance, 'config.instance');
+        assert.uuid(config.server_uuid, 'config.server_uuid');
+    } catch (e) {
+        callback(e);
+        return;
+    }
+
+    callback(null, {});
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/common.js]: ok.

File: lib/common.js [hunk 17]
```diff
@@ -5,82 +5,621 @@
...
+    forkExecWait(
+        {
+            argv: ['/usr/sbin/mount']
+        },
+        function(err, info) {
+            var found = false;
+            var idx;
+            var lines;
+            var matchRe = new RegExp(
+                '^' + INSTRUCTION_ROOT + ' on zones/[0-9a-f-]+/data '
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/common.js]: thanks.

File: lib/common.js [hunk 18]
```diff
@@ -5,82 +5,621 @@
...
+
+            callback(err, {
+                filesWritten: filesWritten,
+                instrsWritten: instrsWritten
+            });
+        }
+    );
+}
+
+function addCounter(counterName, value, labels) {
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/common.js]: thanks.

File: lib/common.js [hunk 19]
```diff
@@ -5,82 +5,621 @@
...
+                }
+            }
+
+            callback(null, found);
+        }
+    );
+}
+
+function writeGarbageInstructions(_self, opts, callback) {
+    assert.object(_self, '_self');
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/common.js]: ok.

File: lib/common.js [hunk 20]
```diff
@@ -5,82 +5,621 @@
...
+                                    next();
+                                });
+                            },
+                            function _mvFileFinal(_, next) {
+                                fs.rename(filenameTemp, filenameFinal, next);
+                            }
+                        ]
+                    },
+                    function _writeFilePipeline(err) {
+                        // XXX err?
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/common.js]: yep.

File: lib/common.js [hunk 21]
```diff
@@ -5,82 +5,621 @@
...
+                        instrsWritten += opts.records[storageId].length;
+                        filesWritten++;
+                        cb(err);
+                    }
+                );
+            },
+            inputs: Object.keys(opts.records)
+        },
+        function _pipelineComplete(err) {
+            // XXX log.error if err?
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/common.js]: ok.

File: bin/garbage-buckets-consumer.js [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+// interfaces added with MANTA-4320.
+//
+// This means calling the `getgcbatch` RPC and writing instructions out to
+// local "instructions" files in:
+//
+//  * /var/spool/manta_gc/<storageId>/*
+//
+// where they will be picked up and transferred to the storage zones for
+// processing. Once records are written out locally, they will be removed from
+// bbuckets-mdapi using the `deletegcbatch` RPC.
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply bin/garbage-buckets-consumer.js]: thanks.

File: bin/garbage-buckets-consumer.js [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+//
+// The Instructions files have:
+//
+//  * storageId
+//  * objectId
+//  * creatorId/ownerId
+//  * shardId
+//  * bytes (for metrics/auditing)
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply bin/garbage-buckets-consumer.js]: yep. Thanks.

File: bin/garbage-buckets-consumer.js [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+                    //
+                    // We create a separate GarbageBucketsConsumer instance for each
+                    // buckets-mdapi shard this GC is assigned to manage.
+                    //
+                    for (
+                        idx = 0;
+                        idx < ctx.config.buckets_shards.length;
+                        idx++
+                    ) {
+                        shard = ctx.config.buckets_shards[idx].host;
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply bin/garbage-buckets-consumer.js]: ok.

File: lib/garbage-buckets-consumer.js [hunk 25]
```diff
@@ -0,0 +1,760 @@
...
+                },
+                'Error reading garbage batch.'
+            );
+
+            callback(err);
+            return;
+        }
+
+        assert.ok(
+            garbageBatch.garbage.length < MAX_GARBAGE_RECORDS,
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/garbage-buckets-consumer.js]: Yeah, I guess so. This is kinda lame. But you're right.

File: bin/garbage-buckets-consumer.js [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+
+                        childLog = logger.child({
+                            component: 'GarbageBucketsConsumer',
+                            shard: shard
+                        });
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply bin/garbage-buckets-consumer.js]: ok.

File: lib/garbage-buckets-consumer.js [hunk 27]
```diff
@@ -0,0 +1,760 @@
...
+            ) {
+                storageId = garbageRecord.sharks[sharkIdx].manta_storage_id;
+
+                if (!records.hasOwnProperty(storageId)) {
+                    records[storageId] = [];
+                }
+
+                // NOTES:
+                //
+                //  Unlike dir-style Manta, buckets not track owner and
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/garbage-buckets-consumer.js]: thanks.

File: lib/garbage-buckets-consumer.js (hunk 1 above)
Comment (joshwilsdon) [reply lib/garbage-buckets-consumer.js]: ok.

File: lib/garbage-uploader.js (hunk 2 above)
Comment (joshwilsdon) [reply lib/garbage-uploader.js]: thanks.

File: boot/setup.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (trentm) [reply boot/setup.sh]: Yes, in manta-scripts.git. I hadn't noticed that deps/manta-scripts wasn't at latest.

**Comment by joshwilsdon** [comment]
@trentm currently the tunables:

 * GC_RECORD_READ_BATCH_SIZE
 * GC_RECORD_READ_WAIT_INTERVAL

are used here by both buckets and dir-style consumers. These values come from manta v1's garbage-collector and I found that they might be set to goofy values (e.g. nightly-2 has GC_RECORD_READ_WAIT_INTERVAL set to 3 seconds). The other problem here is that it's not possible to separate the values for buckets and moray in the case that one or the other is overloaded.

I'm now thinking that perhaps what we should do use use 3 *new* tunables (all still optional):

 * GCv2_DIR_BATCH_SIZE
 * GCv2_DIR_BATCH_INTERVAL_MS
 * GCv2_BUCKETS_BATCH_INTERVAL_MS

Do you have an opinion?

**Comment by trentm** [comment]
> Do you have an opinion?

- +1 to using new vars given your reasons.
- don't use a lowercase "v" in there. Let's keep them all uppercase.  Or perhaps arguably we could not bother with the "v2". I understand that the "v2/V2" in there helps disambiguate.
- Please add a note (could just cut 'n paste) for these vars to https://jira.joyent.us/browse/MANTA-5091 (a ticket to start a reference doc for Manta SAPI metadata vars).

File: bin/garbage-buckets-consumer.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply bin/garbage-buckets-consumer.js]: fwiw: The reason I stay away from `for ... of` and `for ... in` and in fact over time have been trying to move things I'm updating for other reasons away from that, is that several years ago I was told they're significantly slower than a normal `for` loop. A quick test seems to confirm that's still the case with node v6 at least:

```
[root@947f10a3 (garbage-collector) ~]$ cat test.js
var array1 = [];
var array2 = [];
var array3 = [];
var array4 = [];
var idx;
var before;

function elapsedSince(beginning, prev) {
    var elapsed;
    var timeDelta;

    timeDelta = process.hrtime(beginning);
    elapsed = timeDelta[0] + timeDelta[1] / 1e9;

    if (prev) {
        elapsed -= prev;
    }

    return elapsed;
}

before = process.hrtime();
for (idx1 = 0; idx1 < 100000; idx1++) {
    array1.push(idx1);
    array2 = [];
    for (idx2 = 0; idx2 < 100000; idx2++) {
        array2.push(idx2);
    }
}
console.log('normal for elapsed: ' + elapsedSince(before));

before = process.hrtime();
for (let idx1 of array1) {
    array3.push(idx1);
    array4 = [];
    for (let idx2 of array2) {
        array4.push(idx2);
    }
}
console.log('for of elapsed: ' + elapsedSince(before));
[root@947f10a3 (garbage-collector) ~]$ node test.js
normal for elapsed: 628.462260518
for of elapsed: 1186.575101127
[root@947f10a3 (garbage-collector) ~]$
```

I don't suspect that the for loop will be the bottleneck here, but I also don't think the alternative syntax is worth making it ~2x slower.

File: Makefile [hunk 28]
```diff
@@ -55,11 +55,12 @@ BUILDIMAGE_PKGSRC = nginx-1.14.2 # used by tests to approximate mako
 AGENTS = amon config registrar
 
 .PHONY: all
-all: $(STAMP_NODE_PREBUILT) $(STAMP_NODE_MODULES) install
+all: $(STAMP_NODE_PREBUILT) release
```
Comment (trentm) [inline Makefile]: Personally I'd drop `release` from being a dep for `all`. Other components typically don't do that. The Jenkinsfile typical build call is:
```
make print-BRANCH print-STAMP all release publish buildimage bits-upload
```

File: README.md [hunk 29]
```diff
@@ -53,3 +53,22 @@ actual files for the deleted objects. The `garbage-deleter` on the storage zone
 reads these instructions and does the actual deletion before deleting each
 instruction file.
 
+When a buckets object is deleted in Manta the metadata is removed from the
+buckets-mdapi and similar to the Directory-style a record is written that
+indicates an object is read for garbage collection. With buckets-mdapi garbage
```
Comment (trentm) [inline README.md]:
```suggestion
indicates an object is ready for garbage collection. With buckets-mdapi, garbage
```

File: docs/README.md [hunk 30]
```diff
@@ -0,0 +1,249 @@
...
+configuration active. You should be able to confirm this by running:
+
+```
+manta-oneach -s garbage-collector "json dir_shards buckets_shards < /opt/smartdc/manta-garbage-collector/etc/config.json"
+```
+
+At this point garbage collection should be working for all deleted objects in
+both buckets and directory-style Manta.
+
+## Configuring garbage-collector Tunables
```
Comment (trentm) [inline docs/README.md]: Ah, I hadn't realized you were doc'ing these here. That other SAPI config vars doc (for MANTA-5091) can just point here.

File: Makefile (hunk 28 above)
Comment (joshwilsdon) [reply Makefile]: sold.

File: README.md (hunk 29 above)
Comment (joshwilsdon) [reply README.md]: Thanks!

File: docs/README.md (hunk 30 above)
Comment (joshwilsdon) [reply docs/README.md]: ok!

**Pull request merged**